[Table] Add aria-label & caption in table demos

adeelibr opened the pull request:
Closes https://github.com/mui-org/material-ui/issues/17679

<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui-org/material-ui/blob/master/CONTRIBUTING.md#submitting-a-pull-request).

oliviertassinari commented:
Are aria-and caption meant to be used at the same time? Which one should be prefered?

adeelibr commented:
The difference that I was able to find between `<caption/>` tag & `aria-label` attribute is;

- Use caption element to provide an accessible name for a data table.
- Use aria-label attribute to provide an accessible name for a data table (NOTE: inconsistent browser/AT support).

Reference: https://fae.disability.illinois.edu/rulesets/TABLE_2/

So i ended up using both, which I am now thinking is not the right approach maybe. What do you suggest? @oliviertassinari

eps1lon commented:
It's basically the difference between label and description. `caption` works as an implicit reference via `aria-describedby` while `aria-label` is just a label.

If we want to include this we should provide explicit examples not just repeating the words.

adeelibr commented:
> If we want to include this we should provide explicit examples not just repeating the words.

So just `aria-label` or use `caption (with more details)` with `aria-describedby` or both at the same time? 😅

eps1lon commented:
> So just aria-label or use caption (with more details) with aria-describedby or both at the same time? sweat_smile

Can't give you a clear prescription here but at least something more specific than `"label"` and `"label" + caption`.

ahayes91 commented:
> Are _aria-_ and _caption_ meant to be used at the same time? Which one should be preferred? cc @ahayes91

I'm definitely not an expert or regular user of these to be honest. I need to do some more testing, but I tried out `<Table><caption>Some caption</caption>...</Table>` and VoiceOver with Chrome does not announce the caption when using regular tab navigation, whereas `<Table aria-label="Some caption">...</Table>` did. `<Table aria-label="Some caption">...</Table>` will also be announced with table-specific navigation on VoiceOver (see the next table command here) https://dequeuniversity.com/screenreaders/voiceover-keyboard-shortcuts 

I just tried ChromeVox extension with Chrome (MacOS) and it doesn't seem like `<Table aria-label="Some caption">...</Table>` is announced at all using either tab navigation or table mode. 

Will update this comment as I test some more this morning!

adeelibr commented:
I have added an `aria-describedby` along with better captions (I hope 😅)
is it better now? 😅😅

adeelibr commented:
> I've applied CSS to visually hide the <caption> element:
>
> position: fixed;
> bottom: 200vh;

@ahayes91 I have not worked with accessibility that much, but does `display: none` have the same effect in terms of screen-readers like VoiceOver etc as that of `position: fixed; bottom: 200vh` 🤔

ahayes91 commented:
> > I've applied CSS to visually hide the  element:
> > position: fixed;
> > bottom: 200vh;
> 
> @ahayes91 I have not worked with accessibility that much, but does `display: none` have the same effect in terms of screen-readers like VoiceOver etc as that of `position: fixed; bottom: 200vh` 🤔

Quite the opposite 😄 `display: none` will hide the region from screenreaders, same for `visibility: hidden` for most screenreaders. https://webaim.org/techniques/css/invisiblecontent/

adeelibr commented:
thank you for the explanation @ahayes91.

Should I add .hidden class to the `<caption />` element which has `position: fixed; bottom: 200vh` styled apply to it, to hide it from the demo? @eps1lon

oliviertassinari commented:
What do you guys think of using, and only, using caption?
- it's the first approach in https://www.w3.org/WAI/tutorials/tables/caption-summary/
- it's the documented approach in https://getbootstrap.com/docs/4.3/content/tables/#captions

adeelibr commented:
> * https://www.w3.org/WAI/tutorials/tables/caption-summary/

According to this if we are using `<caption />` we don't actually need `aria-describedby` property & this property is only used when

> The element containing the summary doesn’t need to be in front of the table in the document, but it helps users to discover the summary more quickly if the summary is near the table, especially if they are not using a screen reader. i.e,

```html
<p id="tblDesc">Column one has the location and size of accommodation, other columns show the type and number of properties available.</p>
<table aria-describedby="tblDesc">
```

So would a simple table with caption work like below;

```html
<table>
    <caption>lorem ipsum lorem ipsum lorem ipsum</caption>
    <!-- TABLE CONTENT -->
</table>
```

eps1lon commented:
> Should I add .hidden class to the <caption /> element which has position: fixed; bottom: 200vh styled apply to it, to hide it from the demo? @eps1lon

caption should be visible for sighted users.

Either use `caption` or `aria-describedby`. An `aria-labe` is good in any case.

adeelibr commented:
> Either use caption or aria-describedby. An aria-labe is good in any case.

I have made the changes, can you kindly review again. 😄 😅

oliviertassinari commented:
@adeelibr Given the visual implication of the change (Argos-CI diffs). What do you think of:
1. Removing the caption from the demos, keeping the aria-label, only.
2. Add an `Accessibility` section at the bottom of the markdown page, add a demo like https://getbootstrap.com/docs/4.3/content/tables/#captions with the caption element and aria-label.
3. Add some CSS to the Table component to handle the `<caption>` element.

I can help with 3.

adeelibr commented:
on it.

oliviertassinari commented:
@adeelibr Update.

I have tried to handle 3.
Regarding 1. following @eps1lon recommendations, I think that we can replace the caption by aria-described by elements at the bottom of the tables.
2. It would still be useful with some text explaining how to make the table accessible, after all, I'm not sure that we need a new demo for it.

oliviertassinari commented:
Sorry, I have commented too early. After more thoughts, I come to the following conclusion:

1. We need to handle the CSS for the `caption` element. I have pushed a commit in this direction: ddf5401d35e7a46d22132ce443f04b049c301abc.
1. Because the demos, we have, use the "Paper" approach, rendering a caption inside them doesn't look good. I don't think that it's an option we can keep. We need something else (the aria-label is great). 
1. If we are willing to accept the "caption" overhead in all the demos (my personal inclination would be slightly against, but if @eps1lon think that we should, it's fine by me) we would need to replace the caption with a Typography color="secondary" at the bottom, outside the paper, with a aria-describedby link with the table. This is related to point 2.
1. Because accessibility is important, I recommend adding an `Accessibility` documentation section at the bottom of the page. It's something we try to do in all the component pages. We could cover the description part of the problem, linking relevant resources.
1. I think that we need at least one demo with a caption and one with a aria-describedby in the page. If we don't have them, the new proposed accessibility section in 4. would be a great place to host them.

adeelibr commented:
I honestly love it, how 4 lines of CSS can make it so pretty (still amazes me) 😄 😺 But if you don't find it elegant, I can remove the caption tag from all of the tables & only keep aria-label & add a new section for `Accessibility  with 2 demos, 1 with caption & the other with aria-describedby, if that's okay?

attilavago commented:
@ahayes91 this bit might bite us in the behind later...
```
position: fixed;
bottom: 200vh;
```
This is perfect for desktop but on mobile you might end up with unexpected focusing behaviour. It might very well work fine, but worth testing before merging.

eps1lon commented:
The styling looks fine to me. What is the problem currently?

Quick summary as to label, caption and description:

* accessible name derivde from: `aria-labelledby` then `aria-label` then `caption`. First "hit" being prioritized.
* accessible description derived from: `aria-describedby` then `caption`. First "hit" being prioritized.

-- https://w3c.github.io/html-aam/#table-element

eps1lon commented:
>  I don't think that it's something developers want in the majority of the cases (outside of the morality of it).

I'd argue that `aria-label` is perfectly fine if you have a single table per section. For screen reader users `caption` and `aria-label` should be equivalent and for other visual impairments the heading of that section is another visual anchor. 

Captions are a fine thing either way. I hope more people would to add them to their tables since that's what you're supposed to do with your tables. Just like you label your diagrams. Or eat your vitamins :smile:

ahayes91 commented:
> @ahayes91 this bit might bite us in the behind later...
> 
> ```
> position: fixed;
> bottom: 200vh;
> ```
> 
> This is perfect for desktop but on mobile you might end up with unexpected focusing behaviour. It might very well work fine, but worth testing before merging.

Dead right, VoiceOver on iPad focus was going off screen when navigating to the table. Changed the CSS to use `clip` instead and it also made the ChromeVox focus better in table navigation mode:
```
  position: absolute;
  clip: rect(1px, 1px, 1px, 1px);
  border: 0;
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  width: 1px;
  white-space: nowrap;
```

oliviertassinari commented:
@adeelibr Thank you!
